### PR TITLE
Iss811

### LINKED ISSUE #811: New TrackClusterMatcher compatible with KF and GBL
Creating a new TrackClusterMatcher algorithm, named TrackClusterMatcherMinDistance, based on closest distance matching

### PR DESCRIPTION
**Pull request by alspellm**
Added a new TrackClusterMatcherMinDistance.java along with some recon steering-files that use the new matcher.

### CODE REVIEW

**Comment by omar-moreno** [comment]
@alspellm Can you post some basic plots showing residuals etc.?

**Comment by alspellm** [comment]
Residual cuts are hard-coded at the moment:
dx < 20 mm
dy < 20 mm
dt < 8 ns


**Electrons**

image
image
image
image

**Electron Tracks that were NOT matched with a Cluster**
Tracks with P < 0.5 GeV aren't in the acceptance of the Ecal
image








**Positrons**

image
image
image
image


**Positron Tracks that were NOT matched with a Cluster**
Tracks with P < 0.5 aren't in the acceptance of the Ecal
image

File: recon/src/main/java/org/hps/recon/utils/TrackClusterMatcherMinDistance.java [hunk 1]
```diff
@@ -0,0 +1,753 @@
...
+                }
+            }
+            Map.put(track, smallestdrCluster);
+        }
+        return Map;
+    }
+
+    private boolean isTrackInEcal(double trackx, double tracky){
+
+        //Define first order Ecal geometry --> assumed square here,
```
Comment (pbutti) [inline recon/src/main/java/org/hps/recon/utils/TrackClusterMatcherMinDistance.java]: Is it possible to read this numbers from geometry?

File: recon/src/main/java/org/hps/recon/utils/TrackClusterMatcherMinDistance.java [hunk 2]
```diff
@@ -0,0 +1,753 @@
...
+    protected RelationalTable hitToStrips = null;
+    protected RelationalTable trackToData = new BaseRelationalTable(RelationalTable.Mode.ONE_TO_ONE, RelationalTable.Weighting.UNWEIGHTED);
+    protected String trackClusterCollectionName = "GBLTracks";
+    protected String rootFile = String.format("%s_TrackClusterMatching.root",this.trackCollectionName);
+    //Keep false. May be used in future 03.19.21
+    protected Boolean buildParamHistos = false;
+    /*
+     * Cuts used for track+cluster residual matching 
+     */
+    protected double tcut = 8.0;
```
Comment (pbutti) [inline recon/src/main/java/org/hps/recon/utils/TrackClusterMatcherMinDistance.java]: These should be configurable

File: recon/src/main/java/org/hps/recon/utils/TrackClusterMatcherMinDistance.java [hunk 3]
```diff
@@ -0,0 +1,753 @@
...
+import org.lcsim.geometry.subdetector.HPSEcal3;
+import org.lcsim.event.ReconstructedParticle;
+
+
+/**
+ * Utility used to determine if a Tracks and Ecal Cluster are matched.
+ * Class extends AbstractTrackClusterMatcher, which implements
+ * TrackClusterMatcherInter
+ *
+ * @author <a href="mailto:alspellm@ucsc.edu">Alic Spellman</a>
```
Comment (pbutti) [inline recon/src/main/java/org/hps/recon/utils/TrackClusterMatcherMinDistance.java]: Remove

File: recon/src/main/java/org/hps/recon/utils/TrackClusterMatcherMinDistance.java [hunk 4]
```diff
@@ -0,0 +1,753 @@
...
+            Map<Track, Map<Cluster, Double>> trackClusterResidualsMap = new HashMap<Track, Map<Cluster, Double>>(); 
+
+            for(Track track : tracks) {
+
+                 // Derive the charge of the particle from the track.
+                int charge = (int) Math.signum(track.getTrackStates().get(0).getOmega())*flipSign;
+
+                //Track time includes a hard-coded offset
+                double trackt = getTrackTime(track);
+                double tracktOffset = 3.5; //Track time distribution is centered on -4ns, added offset to center ~0
```
Comment (pbutti) [inline recon/src/main/java/org/hps/recon/utils/TrackClusterMatcherMinDistance.java]: Should be configurable

File: recon/src/main/java/org/hps/recon/utils/TrackClusterMatcherMinDistance.java [hunk 5]
```diff
@@ -0,0 +1,753 @@
...
+
+                //Track positions depend on GBL or KF
+                List<Double> trackPos = this.getTrackPositionAtEcal(track);
+                double trackx = trackPos.get(0);
+                double tracky = trackPos.get(1);
+                double trackz = trackPos.get(2);
+
+                //dxoffset to be determined using truth information 03.19.21
```
Comment (pbutti) [inline recon/src/main/java/org/hps/recon/utils/TrackClusterMatcherMinDistance.java]: Same

**Review by pbutti — COMMENTED** [review]
Why 2 steering files. If just for example, just do a minimal one.

File: recon/src/main/java/org/hps/recon/utils/TrackClusterMatcherMinDistance.java (hunk 1 above)
Comment (alspellm) [reply recon/src/main/java/org/hps/recon/utils/TrackClusterMatcherMinDistance.java]: Not sure how to do this. This really isn't necessary to keep probably, I think I should remove. It was only really used as a sanity check

**Comment by normangraf** [comment]
How does this matcher compare to the legacy matcher? It would be nice to see some plots for both MC and data.

File: recon/src/main/java/org/hps/recon/utils/TrackClusterMatcherMinDistance.java (hunk 5 above)
Comment (alspellm) [reply recon/src/main/java/org/hps/recon/utils/TrackClusterMatcherMinDistance.java]: These are the offsets that were originally hard-coded that you asked me to just remove for now. Do you want these to be configurable from the steering-file for the time being?

File: recon/src/main/java/org/hps/recon/utils/TrackClusterMatcherMinDistance.java (hunk 2 above)
Comment (alspellm) [reply recon/src/main/java/org/hps/recon/utils/TrackClusterMatcherMinDistance.java]: Can I add the dx and dy cuts in to the StandardCuts.java file?That seems like a reasonable way to pass these

Comment (pbutti) [reply recon/src/main/java/org/hps/recon/utils/TrackClusterMatcherMinDistance.java]: I find that class confusing. Avoid. Here is OK.

File: recon/src/main/java/org/hps/recon/utils/TrackClusterMatcherMinDistance.java (hunk 5 above)
Comment (pbutti) [reply recon/src/main/java/org/hps/recon/utils/TrackClusterMatcherMinDistance.java]: Perhaps tracktOffset should be a member of the class more than a local variable, so then it's easy to configure from steering file.

File: recon/src/main/java/org/hps/recon/utils/TrackClusterMatcherMinDistance.java (hunk 1 above)
Comment (pbutti) [reply recon/src/main/java/org/hps/recon/utils/TrackClusterMatcherMinDistance.java]: Perhaps isTrackInEcal should be a method of TrackUtils.

**Review by pbutti — APPROVED** [review]

**Review by cbravo135 — APPROVED** [review]
This code is mostly orthogonal to what exists, making the review more simple. It does change the NSigmaMatcher so I did locally run the tests and verify nothing was broken, though not expected from the changes made. Tests all pass, and since this is adding functionality and accepting this PR does not imply the analysis wg has accepted using this matching algorithm for use in any analysis, it is better to take this code now so we can collaboratively study this algorithm as a collaboration for use in a future analysis. Improvements and adjustments can still be made to this before it is used to officially reconstruct anything. Studies should be done before deciding to put this into the prod steering files, and since this PR does not, it appears safe.

**Comment by pbutti** [comment]
Merged

**Pull request merged**